core/org.wso2.carbon.core/src/main/java/org/wso2/carbon/core/internal/CarbonCoreActivator.java

```diff
@@ -69,13 +70,43 @@ public void start(BundleContext context) throws Exception {
                  System.getProperty("user.language") + "-" + System.getProperty("user.country") +
                  ", " + System.getProperty("user.timezone"));
 
-        Security.addProvider(new BouncyCastleProvider());
+        String cryptoProviderIdentifier = getPreferredJceProviderIdentifier();
+        String cryptoProviderClass = getPreferredJceProviderClass(cryptoProviderIdentifier);
+        Security.addProvider((Provider) Class.forName(cryptoProviderClass).getDeclaredConstructor().newInstance());
         if(log.isDebugEnabled()){
-            log.debug("BouncyCastle security provider is successfully registered in JVM.");
+            log.debug(cryptoProviderClass + " security provider is successfully registered in JVM.");
         }
     }
 
     public void stop(BundleContext context) throws Exception {
         dataHolder.setBundleContext(null);
     }
+
+    /**
+     * This method returns the preferred JCE provider identifier to be used.
+     *
+     * @return jce provider identifier name
+     */
+    private static String getPreferredJceProviderIdentifier() {
+        String provider = System.getProperty(ServerConstants.JCE_PROVIDER_PARAMETER);
+        if (ServerConstants.BOUNCY_CASTLE_FIPS_PROVIDER_IDENTIFIER.equalsIgnoreCase(provider)) {
+            return ServerConstants.BOUNCY_CASTLE_FIPS_PROVIDER_IDENTIFIER;
+        }
+        return ServerConstants.BOUNCY_CASTLE_PROVIDER_IDENTIFIER;
+    }
+
+    /**
+     * This method returns the preferred JCE provider class to be used.
+     *
+     * @return jce provider identifier name
+     */
+    private static String getPreferredJceProviderClass(String providerIdentifier) throws Exception {
+        if (ServerConstants.BOUNCY_CASTLE_PROVIDER_IDENTIFIER.equalsIgnoreCase(providerIdentifier)) {
+            return ServerConstants.BOUNCY_CASTLE_PROVIDER_CLASS;
+        } else if (ServerConstants.BOUNCY_CASTLE_FIPS_PROVIDER_IDENTIFIER.equalsIgnoreCase(providerIdentifier)) {
+            return ServerConstants.BOUNCY_CASTLE_FIPS_PROVIDER_CLASS;
+        } else {
+            throw new Exception("Unsupported JCE provider: " + providerIdentifier);
```

Comment:
It would be better to use a more specific exception type, such as IllegalArgumentException, to clearly indicate the cause of the error.

Reply:
Added with https://github.com/wso2/carbon-kernel/pull/3550/commits/24668eb3412c1db0392c5428d6c2bddd2822a1c1